### ISSUE TITLE
[Enhancement] - Async/Autoloadable VFS Packages

### ISSUE THREAD

**Issue by anpylar** [brython-dev/brython#772]
A pull-request in this case seems to make little sense, given that some decisions have to probably be made, before inserting/accepting this. It can be provided as a pull-request.

**Background**

During the development of *AnPyLar*, the need to have packages which can be loaded separately from `brython.js` in an *asynchronous* way was envisioned, including also the possibility of having them packaged in the same file as `brython.js` and `brython_stdlib.js`.

It was possible to devise such a method without patching `Brython` itself, although due to the inner workings of the `brython()` call, there is hack for the insertion of the packages in `$B.path` (to avoid the packages being removed)

**Part 1 The package structure**

The packages are no longer a *JSON* structure inside a variable. They become a `function` which will insert the `JSON` structure in an array for later loading. The format

```
;(function() {
    var vfspath = "anpylar.vfs.js"
    var $vfs = "...."  // JSON content of an VFS package

    if(window.__ANPYLAR__ === undefined)
        window.__ANPYLAR__ = {autoload: []}  // ensure global scope

    window.__ANPYLAR__.autoload.push(function($B) {
        // brython removes 1st 2 entries later if stdlib import is true
        // our entry would be removed if placed 1st
        $B.path.splice(2, 0, vfspath)
        $B.imported['_importlib'].VFSAutoPathFinder(vfspath, $vfs)  // autoload
    })
})()
```

The references to `__ANPYLAR__` would obviously be replaced with references to `__BRYTHON__`. The code takes care of looking for the object `__ANPYLAR__` and if not yet defined, it does create it with just the `autoload` array. This is compatible with the code that kick-starts the creation of `__BRYTHON__` which does

```
var __BRYTHON__=__BRYTHON__ ||{}
```

As you may see above, the package auto-inserts itself in `$B.path` (i.e.: `__BRYTHON__.path`) taking care of the side effects that may happen inside `brython()`, and it also adds an entry into `$B.imported` with a  `VFSAutoPathFinder` instance (see below)

**Part 2 Before `brython()` really acts**

This code would be at the top of the `brython()`  call

```
    // define __ANPYLAR__ if not defined
    if(window.__ANPYLAR__ === undefined)
        window.__ANPYLAR__ = {autoload: []}

    // autoload packages if any has registered
    var autoload = window.__ANPYLAR__.autoload
    if(autoload !== undefined)
        for(var i=0; i < autoload.length; i++)
            autoload[i]($B)
```

Just before any other operation, the existence of the `autoload` array (and the `__ANPYLAR__`  object also, although this wouldn't be needed) is checked. Once an array is in place it is traversed to auto-load the packages.

Because autoload packages are pure Javascript and are embedded in an auto-executable function, they can be loaded in separate files using a standard `<script type="text/javascript" src="..." async></script>` tag. Or as it is the case with *AnPyLar*, embedded along with `brython.js`

The code above for `brython()` would only execute when all the scripts have been loaded. As longs as the scripts are not added dynamically, order is preserved. In the case of scripts in the `<HEAD>` element, scripts can be put in any order, because they will all be downloaded before `DOMContentLoaded` or `readyState == complete` are `true`.

If the scripts (and `brython.js` itself) are in the `<BODY>` packages have to be there before `brython.js`.

**Part 3 VFSAutoPathFinder**

This is simply a slight modification of  `VFSPathFinder`, which doesn't download the package, because its presence is guaranteed. The code for auto-generation (probably in `py_import.js`:

```
    ;(function($B){
        var _b_ = $B.builtins

        function parent_package(mod_name) {
            var parts = mod_name.split('.')
            parts.pop()
            return parts.join('.')
        }

        function sfs_hook(path, vfs) {
            if (path.substr(-1) == '/')
                path = path.slice(0, -1)

            // it doesn't really make sense to check for a extension, because we
            // know the code is wrapped in a variable which is available
            var ext = path.substr(-7)
            if(ext != '.vfs.js')
                throw _b_.ImportError('VFS file URL must end with .vfs.js extension')

            s = {__class__: sfs_hook.$dict, path: path}
            sfs_hook.$dict.load_vfs(s, vfs)
            return s
        }

        sfs_hook.__class__ = $B.$factory

        sfs_hook.$dict = {
            $factory: sfs_hook,
            __class__: $B.$type,
            __name__: 'VFSAutoPathFinder',

            load_vfs: function(s, vfs) {
                s.vfs = vfs
                $B.path_importer_cache[s.path + '/'] = s
            },
            find_spec: function(s, fullname, module) {
                var stored = s.vfs[fullname]
                if (stored === undefined)
                    return _b_.None

                var is_package = stored[2]
                return {
                    name : fullname,
                    // newspec function in original does simply __class__ ...
                    __class__: $B.$ModuleDict,
                    loader: $B.imported['_importlib'].VFSFinder,  // finder_VFS
                    // FIXME : Better origin string.
                    origin: s.path + '#' + fullname,
                    // FIXME: Namespace packages ?
                    submodule_search_locations: is_package ? [s.path] : _b_.None,
                    loader_state: {stored: stored},
                    // FIXME : Where exactly compiled module is stored ?
                    cached: _b_.None,
                    parent: is_package ? fullname : parent_package(fullname),
                    has_location: _b_.True
                }
            },
            invalidate_caches: function(s) {
            }
        }

        sfs_hook.$dict.__mro__ = [_b_.object.$dict]

        $B.imported['_importlib'].VFSAutoPathFinder = sfs_hook
    })($B)
```

There are some additional topics which have been implemented in *AnPyLar* which may be added to `brython` or not.

**Option 1 - Auto Starting**

*AnPyLar* does auto-start `brython()` with this code

```
// BRYTHON must already be in place
var $al = function() {anpylar_load(__BRYTHON__)}
if(document.readyState === 'interactive' || document.readyState === 'complete')
    $al()  // if async ... the DOM will probably be there
else
    window.addEventListener('DOMContentLoaded', $al, true)
```

Where `anpylar_load` is made up of the parts seen above and finally calls `brython()`. The key thing here is that once a user has added a link to load `brython.js`, one could *"safely*" assume that he/she wants to have it running.

  - Why not doing it automatically?

**Option 2 - Auto Script Start**

This code is used in *AnPyLar* to automatically look for a script or generate one.

```
    var $scripts = document.getElementsByTagName('script'),
        slen = $scripts.length

    for(var i=0, found=false; i < slen && !found; i++) {
        var t = $scripts[i].type
        if(t == 'text/python' || t == 'text/python3')
            found = true
    }
    if(!found) {
        var script = document.createElement("script")
        script.type = 'text/python'
        script.innerHTML = 'import app; app.AppModule()'
        document.getElementsByTagName('head')[0].appendChild(script)
```

The idea is:

  - If the end user has provided a script: **Use it**
  - If not: generate one that follows a convention (the actual convention can be discussed): `import app; app.AppModule()` is simply the one chosen in `AnPyLar`

The goal being: make it as simple as possible for the end user. Offer the end user a default auto-start convention and at the same time, provide freedom to have custom scripts.

Please have a look.

**Comment by PierreQuentel** [comment]
I don't forget this request, but I was working on a major rewriting of the way Brython scripts are run, which includes pre-loading of imported modules before executing a script and using a cache in an indexedDB database. These operations will be done asynchronously, and the changes will undoubtedly impact the code that your proposal addresses.

The first results I got are very promising in terms of speed of imports, which is one of the major problems today.

Unfortunately I had to postpone work on this feature because I found a major bug while working on issue #766. When this bug is fixed I will resume work on the async imports and hopefully it will be published at the end of next week.

**Comment by anpylar** [comment]
No issues Pierre. These are proposals from things that were needed, or seen as needed, for a specific development and which have been deemed as potentially useful for *Brython* itself.

If some of the things can still be applied after your re-design is mature, good. If some of then could be applied with some changes, also good. And if some of them have to be ditched, because something better has come along or they no longer fit, so be it.

Eagerly awaiting the new functionality!

**Comment by PierreQuentel** [comment]
This is a very old issue, and many changes have been introduced since, including improvements to the indexedDB cache and a [simple way to insert importable modules](https://groups.google.com/forum/?fromgroups=#!topic/brython/EfBUmkrdgXA) in a page, so I close this issue.
